src/lib/transition_frontier/frontier_base/frontier_hash.ml

```diff
@@ -0,0 +1,123 @@
+open Core_kernel
+open Coda_base
+open Coda_transition
+open Signature_lib
+open Module_version
+open Digestif.SHA256
+
+module Stable = struct
+  module V1 = struct
+    module T = struct
+      type t = Digestif.SHA256.t [@@deriving version {asserted}]
+
+      let to_yojson hash = [%derive.to_yojson: string] (to_raw_string hash)
+
+      let of_yojson json =
+        let open Result.Let_syntax in
+        let%bind str = [%derive.of_yojson: string] json in
+        match of_raw_string_opt str with
+        | Some hash ->
+            Ok hash
+        | None ->
+            Error "invalid raw hash"
+
+      include Binable.Of_stringable (struct
+        type nonrec t = t
+
+        let of_string = of_hex
+
+        let to_string = to_hex
+      end)
+    end
+
+    include T
+    include Registration.Make_latest_version (T)
+  end
+
+  module Latest = V1
+
+  module Module_decl = struct
+    let name = "transition_frontier_hash"
+
+    type latest = Latest.t
+  end
+
+  module Registrar = Registration.Make (Module_decl)
+  module Registered_V1 = Registrar.Register (V1)
+end
+
+include Stable.Latest
+
+type transition = {source: t; target: t}
+
+let equal t1 t2 = equal t1 t2
+
+let empty = digest_string ""
+
+let merge_string t1 string = digestv_string [to_hex t1; string]
+
+let to_string = to_hex
+
+let merge_state_hash acc state_hash =
+  merge_string acc (State_hash.raw_hash_bytes state_hash)
+
+(* This currently only hashes the creator of the staged ledger
+ * diff. The previous hash is already encapsulated in the
+ * transition frontier incremental diff hash via the state
+ * hash since the Blockchain_state contains the previous
+ * staged ledger hash already. The only information missing
+ * from this is the actual contents of the diff itself
+ * (the transactions). This can't be included in O(1) as is
+ * since it is not precomputed, so it is left out here for now. *)
+let merge_staged_ledger_diff acc diff =
+  (* TODO: hash target ledger hash? (does this need to be computed?) *)
+  merge_string acc
+    (Public_key.Compressed.to_string (Staged_ledger_diff.creator diff))
+
+let merge_transition acc transition =
```

Comment:
The staged ledger hash inside the transition is merely a commitment to the staged ledger. Since the snark doesn't say anything about it, it's completely possible for someone to build a block which has a different staged ledger than the hash points to. That said, it may not be important to capture that directly in the frontier hash as we only use it for synchronization and not security, and at the point we actually check it, we should have already validated the staged ledger hash from the block.